File: watch/wearos/src/main/java/app/vescape/wear/MirrorState.kt
package app.vescape.wear

const val WATCH_FRAME_INTERVAL_MS = 250L
const val MIRROR_DISCONNECTED_TIMEOUT_MS = WATCH_FRAME_INTERVAL_MS * 3

enum class MirrorStatus {
    LIVE,
    STALE,

    /** Legacy phone frame; retained for compatibility with older phone builds. */
    WAITING,

    /** No fresh frames at all — see [PhoneLink] for why. */
    DISCONNECTED,
}

/**
 * Watch-local view of the phone link, derived by [PhoneLinkMonitor] from `NodeClient` +
 * `CapabilityClient`. Only meaningful while no frames arrive — it names the reason for the wait.
 */
enum class PhoneLink {
    UNKNOWN,

    /** No connected Wear node at all: Bluetooth link to the phone is down. */
    NO_PHONE,

    /** A phone is connected but the Vescape app capability is absent — app missing or too old. */
    PHONE_ONLY,

    /** The Vescape phone app is installed and the node is reachable; it just isn't pushing. */
    APP_REACHABLE,
}

data class MirrorState(
    val status: MirrorStatus,
    val frame: WatchFrame?,
)

object MirrorStateReducer {
    fun reduce(
        frame: WatchFrame?,
        lastFrameAtMs: Long?,
        nowMs: Long,
        timeoutMs: Long = MIRROR_DISCONNECTED_TIMEOUT_MS,
    ): MirrorState {
        if (frame == null || lastFrameAtMs == null || nowMs - lastFrameAtMs > timeoutMs) {
            return MirrorState(MirrorStatus.DISCONNECTED, null)
        }

        if (frame.waiting) {
            return MirrorState(
                MirrorStatus.WAITING,
                frame.copy(
                    speed = null,
                    duty = null,
                    battery = null,
                    motorTemp = null,
                    ctrlTemp = null,
                ),
            )
        }

        return MirrorState(
            status = if (frame.stale) MirrorStatus.STALE else MirrorStatus.LIVE,
            frame = frame,
        )
    }
}

File: modules/vescape-core/ios/watch/MirrorState.swift
import Foundation

/// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorState.kt `MirrorStatus`
enum MirrorStatus {
  case live
  case stale

  /// Legacy phone frame; retained for compatibility with older phone builds. No iOS phone sets the
  /// flag, so this is unreachable on watchOS today — it exists because the reducer is the contract,
  /// and a contract that silently drops a flag the wire format defines is not one.
  case waiting

  /// No fresh frames at all — see `MirrorPhoneLink` for why.
  case disconnected
}

/// Watch-local view of the phone link. Only meaningful while no frames arrive — it names the reason
/// for the wait. Android derives it from `NodeClient` + `CapabilityClient`; watchOS reads the same
/// three facts off `WCSession` (paired, companion installed, reachable).
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorState.kt `PhoneLink`
enum MirrorPhoneLink {
  case unknown

  /// No paired phone at all: the watch is not talking to an iPhone.
  case noPhone

  /// A phone is paired but the Vescape companion is absent — app missing or too old.
  case phoneOnly

  /// The Vescape phone app is installed and reachable; it just isn't pushing.
  case appReachable
}

/// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorState.kt `MirrorState`
struct MirrorState: Equatable {
  let status: MirrorStatus
  let frame: WatchFrame?

  init(status: MirrorStatus, frame: WatchFrame?) {
    self.status = status
    self.frame = frame
  }
}

/// Pure frame-plus-clock -> Mirror State reduction. The wrist ages a stopped stream into
/// `disconnected` off its own clock, so a phone that stops pushing without saying so still shows as
/// offline rather than freezing on a plausible-looking number.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorState.kt `MirrorStateReducer`
enum MirrorStateReducer {
  /// Cadence assumed until two frames have been seen; matches the phone's default push interval.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorState.kt `WATCH_FRAME_INTERVAL_MS`
  static let frameIntervalMs: Int64 = 250

  /// Three missed frames means disconnected. The phone's push cadence is a rider setting, so the
  /// window is measured from the frames that actually arrive rather than assumed — a hardcoded
  /// window pins the mirror to `disconnected` on every cadence above the default. Clamped so
  /// neither a burst nor a long stall distorts it.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorState.kt `MIRROR_DISCONNECTED_MIN_TIMEOUT_MS`
  static let minTimeoutMs: Int64 = 750
  static let maxTimeoutMs: Int64 = 30_000

  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MirrorState.kt `mirrorDisconnectedTimeoutMs`
  static func disconnectedTimeoutMs(frameGapMs: Int64?) -> Int64 {
    min(max((frameGapMs ?? frameIntervalMs) * 3, minTimeoutMs), maxTimeoutMs)
  }

  static func reduce(
    frame: WatchFrame?,
    lastFrameAtMs: Int64?,
    nowMs: Int64,
    timeoutMs: Int64 = disconnectedTimeoutMs(frameGapMs: nil)
  ) -> MirrorState {
    guard let frame, let lastFrameAtMs, nowMs - lastFrameAtMs <= timeoutMs else {
      return MirrorState(status: .disconnected, frame: nil)
    }

    if frame.waiting {
      // The lanes carry no data in a waiting frame and must not be rendered: emptied here rather
      // than trusted to every readout to check the flag.
      var blanked = frame
      blanked.speed = nil
      blanked.duty = nil
      blanked.battery = nil
      blanked.motorTemp = nil
      blanked.ctrlTemp = nil
      return MirrorState(status: .waiting, frame: blanked)
    }

    return MirrorState(status: frame.stale ? .stale : .live, frame: frame)
  }
}
